null pointer can occur when reconciling routes

<!-- If you need to report a security issue with kf, send an email to kf-security@google.com. -->

## What version of kf?

<!-- Paste the output of kf version -->

## Expected Behavior

<!-- Briefly describe what you expect to happen -->
Deleting an app should not result in a NPE on the kf controller.

## Actual Behavior

<!-- Briefly describe what is actually happening -->
Deleting an app resulted in a NPE on the kf controller.
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13bdb97]

goroutine 273 [running]:
github.com/google/kf/pkg/apis/kf/v1alpha1.(*Route).SetDefaults(...)
        /usr/local/google/home/mattysweeps/src/kf/pkg/apis/kf/v1alpha1/route_defaults.go:50
github.com/google/kf/pkg/reconciler/route.(*Reconciler).ApplyChanges(0xc000311410, 0x1aa52a0, 0xc0008d9dd0, 0x0, 0xc000635b01, 0xc000cc1c18, 0x1, 0x0)
        /usr/local/google/home/mattysweeps/src/kf/pkg/reconciler/route/reconciler.go:101 +0x37
github.com/google/kf/pkg/reconciler/route.(*Reconciler).reconcileRoute(0xc000311410, 0x1aa52a0, 0xc0008d9dd0, 0xc00003f270, 0x4, 0xc00003f275, 0x3f, 0xc000cc1c18, 0xc0008d9dd0, 0x15adf00)
        /usr/local/google/home/mattysweeps/src/kf/pkg/reconciler/route/reconciler.go:90 +0x157
github.com/google/kf/pkg/reconciler/route.(*Reconciler).Reconcile(0xc000311410, 0x1aa52a0, 0xc0008d9dd0, 0xc00003f270, 0x44, 0xc000cc1c18, 0x1aa52a0)
        /usr/local/google/home/mattysweeps/src/kf/pkg/reconciler/route/reconciler.go:61 +0x10e
knative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc0001d8bc0, 0x1856d00)
        /usr/local/google/home/mattysweeps/go/pkg/mod/knative.dev/pkg@v0.0.0-20190626215608-1104d6c75533/controller/controller.go:330 +0x514
knative.dev/pkg/controller.(*Impl).Run.func1(0xc000976570, 0xc0001d8bc0)
        /usr/local/google/home/mattysweeps/go/pkg/mod/knative.dev/pkg@v0.0.0-20190626215608-1104d6c75533/controller/controller.go:282 +0x51
created by knative.dev/pkg/controller.(*Impl).Run
        /usr/local/google/home/mattysweeps/go/pkg/mod/knative.dev/pkg@v0.0.0-20190626215608-1104d6c75533/controller/controller.go:280 +0x1a5
```
This happens because `original` is returned nil.
https://github.com/google/kf/blob/master/pkg/reconciler/route/reconciler.go#L77

## Steps to Reproduce the Problem

<!-- How can a maintainer reproduce this issue (be detailed) -->
I've only seen this once, and I'm pretty sure it's when I deleted an app. Note that the app did in fact get deleted once the controller rebooted.

--- a/pkg/reconciler/app/resources/route.go
+++ b/pkg/reconciler/app/resources/route.go
@@ -70,6 +70,15 @@ func mustRequirement(key string, op selection.Operator, val string) labels.Requi
 }
 
 // MakeRouteSelector creates a labels.Selector for listing all the
+// corresponding Routes excluding Path.
+func MakeRouteSelectorNoPath(spec v1alpha1.RouteSpecFields) labels.Selector {
+	return labels.NewSelector().Add(
+		mustRequirement(v1alpha1.RouteHostname, selection.Equals, spec.Hostname),
+		mustRequirement(v1alpha1.RouteDomain, selection.Equals, spec.Domain),
+	)
+}
+
+// MakeRouteSelector creates a labels.Selector for listing all the
 // corresponding Routes.
 func MakeRouteSelector(spec v1alpha1.RouteSpecFields) labels.Selector {
 	return labels.NewSelector().Add(

--- a/pkg/reconciler/base.go
+++ b/pkg/reconciler/base.go
@@ -60,7 +60,9 @@ type Base struct {
 	// the expense of slightly greater verbosity.
 	Logger *zap.SugaredLogger
 
-	namespaceLister v1listers.NamespaceLister
+	// NamespaceLister allows us to list Namespaces. We use this to check for
+	// terminating namespaces.
+	NamespaceLister v1listers.NamespaceLister
 }
 
 // NewBase instantiates a new instance of Base implementing
@@ -81,7 +83,7 @@ func NewBase(ctx context.Context, controllerAgentName string, cmw configmap.Watc
 		ConfigMapWatcher: cmw,
 		Logger:           logger,
 
-		namespaceLister: nsInformer.Lister(),
+		NamespaceLister: nsInformer.Lister(),
 	}
 
 	return base
@@ -90,7 +92,7 @@ func NewBase(ctx context.Context, controllerAgentName string, cmw configmap.Watc
 // IsNamespaceTerminating returns true if the namespace is marked as terminating
 // and false if the state is unknown or not terminating.
 func (base *Base) IsNamespaceTerminating(namespace string) bool {
-	ns, err := base.namespaceLister.Get(namespace)
+	ns, err := base.NamespaceLister.Get(namespace)
 	if err != nil || ns == nil {
 		return false
 	}

--- a/pkg/reconciler/route/controller.go
+++ b/pkg/reconciler/route/controller.go
@@ -17,13 +17,14 @@ package route
 import (
 	"context"
 
-	"github.com/google/kf/pkg/apis/kf/v1alpha1"
+	v1alpha1 "github.com/google/kf/pkg/apis/kf/v1alpha1"
 	routeinformer "github.com/google/kf/pkg/client/injection/informers/kf/v1alpha1/route"
 	"github.com/google/kf/pkg/reconciler"
-	virtualserviceinformer "knative.dev/pkg/client/injection/informers/istio/v1alpha3/virtualservice"
-
+	appresources "github.com/google/kf/pkg/reconciler/app/resources"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/tools/cache"
-
+	networking "knative.dev/pkg/apis/istio/v1alpha3"
+	virtualserviceinformer "knative.dev/pkg/client/injection/informers/istio/v1alpha3/virtualservice"
 	"knative.dev/pkg/configmap"
 	"knative.dev/pkg/controller"
 	"knative.dev/pkg/logging"
@@ -51,10 +52,58 @@ func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl 
 	// Watch for changes in sub-resources so we can sync accordingly
 	routeInformer.Informer().AddEventHandler(controller.HandleAll(impl.Enqueue))
 
+	// Watch for any changes to VirtualServices in the kf namespace.
 	vsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
-		FilterFunc: controller.Filter(v1alpha1.SchemeGroupVersion.WithKind("Route")),
-		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
+		FilterFunc: FilterVSWithNamespace(v1alpha1.KfNamespace),
+		Handler:    controller.HandleAll(EnqueueRoutesOfVirtualService(impl, c)),
 	})
 
 	return impl
+}
+
+// FilterVSWithNamespace makes it simple to create FilterFunc's for use with
+// cache.FilteringResourceEventHandler that filter based on a namespace and if
+// the type is a VirtualService.
+func FilterVSWithNamespace(namespace string) func(obj interface{}) bool {
+	return func(obj interface{}) bool {
+		if object, ok := obj.(metav1.Object); ok {
+			if namespace == object.GetNamespace() {
+				_, ok := obj.(*networking.VirtualService)
+				return ok
+			}
+		}
+		return false
+	}
+}
+
+// EnqueueRoutesOfVirtualService will find the corresponding routes for the
+// VirtualService.  It will Enqueue a key for each one. We aren't able to use
+// EnqueueControllerOf (as other components do), because a VirtualService is
+// NOT owned by a single Route. Therefore, when one changes, we need to grab
+// the collection of corresponding Routes.
+func EnqueueRoutesOfVirtualService(
+	c *controller.Impl,
+	r *Reconciler,
+) func(obj interface{}) {
+	return func(obj interface{}) {
+		vs, ok := obj.(*networking.VirtualService)
+		if !ok {
+			return
+		}
+
+		routes, err := r.routeLister.
+			Routes(vs.Annotations["space"]).
+			List(appresources.MakeRouteSelectorNoPath(v1alpha1.RouteSpecFields{
+				Domain:   vs.Annotations["domain"],
+				Hostname: vs.Annotations["hostname"],
+			}))
+		if err != nil {
+			r.Logger.Warnf("failed to list corresponding routes: %s", err)
+			return
+		}
+
+		for _, route := range routes {
+			c.Enqueue(route)
+		}
+	}
 }

--- a/pkg/reconciler/route/reconciler.go
+++ b/pkg/reconciler/route/reconciler.go
@@ -67,19 +67,19 @@ func (r *Reconciler) reconcileRoute(
 	name string,
 	logger *zap.SugaredLogger,
 ) (err error) {
-	var deleted bool
 	original, err := r.routeLister.
 		Routes(namespace).
 		Get(name)
 	switch {
 	case apierrs.IsNotFound(err):
 		logger.Errorf("Route %q no longer exists\n", name)
-		deleted = true
+		return nil
 
 	case err != nil:
 		return err
 
 	case original.GetDeletionTimestamp() != nil:
+		logger.Errorf("Route %q is being deleted\n", name)
 		return nil
 	}
 
@@ -92,7 +92,7 @@ func (r *Reconciler) reconcileRoute(
 	toReconcile := original.DeepCopy()
 
 	// Reconcile this copy of the route.
-	return r.ApplyChanges(ctx, toReconcile, deleted, logger)
+	return r.ApplyChanges(ctx, toReconcile, logger)
 }
 
 // ApplyChanges updates the linked resources in the cluster with the current
@@ -100,7 +100,6 @@ func (r *Reconciler) reconcileRoute(
 func (r *Reconciler) ApplyChanges(
 	ctx context.Context,
 	origRoute *v1alpha1.Route,
-	deleted bool,
 	logger *zap.SugaredLogger,
 ) error {
 	origRoute.SetDefaults(ctx)
@@ -134,10 +133,11 @@ func (r *Reconciler) ApplyChanges(
 			}
 		} else if err != nil {
 			return err
+		} else if actual.GetDeletionTimestamp() != nil {
+
 		} else if actual, err = r.reconcile(
 			desired,
 			actual,
-			deleted,
 			logger,
 		); err != nil {
 			return err
@@ -147,7 +147,11 @@ func (r *Reconciler) ApplyChanges(
 	return nil
 }
 
-func (r *Reconciler) reconcile(desired, actual *networking.VirtualService, deleted bool, logger *zap.SugaredLogger) (*networking.VirtualService, error) {
+func (r *Reconciler) reconcile(
+	desired *networking.VirtualService,
+	actual *networking.VirtualService,
+	logger *zap.SugaredLogger,
+) (*networking.VirtualService, error) {
 	// Check for differences, if none we don't need to reconcile.
 	semanticEqual := equality.Semantic.DeepEqual(desired.ObjectMeta.Labels, actual.ObjectMeta.Labels)
 	semanticEqual = semanticEqual && equality.Semantic.DeepEqual(desired.Spec, actual.Spec)
@@ -167,30 +171,19 @@ func (r *Reconciler) reconcile(desired, actual *networking.VirtualService, delet
 	existing.ObjectMeta.Labels = desired.ObjectMeta.Labels
 	existing.ObjectMeta.Annotations = desired.ObjectMeta.Annotations
 
-	if deleted {
-		existing.OwnerReferences = algorithms.Delete(
-			v1alpha1.OwnerReferences(existing.OwnerReferences),
-			v1alpha1.OwnerReferences(desired.OwnerReferences),
-		).(v1alpha1.OwnerReferences)
+	// Merge new OwnerReferences and HTTPRoutes
+	existing.OwnerReferences = algorithms.Merge(
+		v1alpha1.OwnerReferences(existing.OwnerReferences),
+		v1alpha1.OwnerReferences(desired.OwnerReferences),
+	).(v1alpha1.OwnerReferences)
 
-		existing.Spec.HTTP = algorithms.Delete(
-			v1alpha1.HTTPRoutes(existing.Spec.HTTP),
-			v1alpha1.HTTPRoutes(desired.Spec.HTTP),
-		).(v1alpha1.HTTPRoutes)
-	} else {
-		existing.OwnerReferences = algorithms.Merge(
-			v1alpha1.OwnerReferences(existing.OwnerReferences),
-			v1alpha1.OwnerReferences(desired.OwnerReferences),
-		).(v1alpha1.OwnerReferences)
+	existing.Spec.HTTP = algorithms.Merge(
+		v1alpha1.HTTPRoutes(existing.Spec.HTTP),
+		v1alpha1.HTTPRoutes(desired.Spec.HTTP),
+	).(v1alpha1.HTTPRoutes)
 
-		existing.Spec.HTTP = algorithms.Merge(
-			v1alpha1.HTTPRoutes(existing.Spec.HTTP),
-			v1alpha1.HTTPRoutes(desired.Spec.HTTP),
-		).(v1alpha1.HTTPRoutes)
-
-		// Sort by reverse to defer to the longest matchers.
-		sort.Sort(sort.Reverse(v1alpha1.HTTPRoutes(existing.Spec.HTTP)))
-	}
+	// Sort by reverse to defer to the longest matchers.
+	sort.Sort(sort.Reverse(v1alpha1.HTTPRoutes(existing.Spec.HTTP)))
 
 	return r.SharedClientSet.
 		Networking().

--- a/pkg/reconciler/route/resources/virtual_service.go
+++ b/pkg/reconciler/route/resources/virtual_service.go
@@ -21,9 +21,9 @@ import (
 	"path"
 
 	"github.com/google/kf/pkg/apis/kf/v1alpha1"
+	"github.com/google/kf/pkg/kf/algorithms"
 	"github.com/gorilla/mux"
 	"github.com/knative/serving/pkg/network"
-	"github.com/knative/serving/pkg/resources"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	istio "knative.dev/pkg/apis/istio/common/v1alpha1"
 	networking "knative.dev/pkg/apis/istio/v1alpha3"
@@ -56,35 +56,41 @@ func MakeVirtualService(routes []*v1alpha1.Route) (*networking.VirtualService, e
 	namespace := routes[0].Namespace
 	hostname := routes[0].Spec.RouteSpecFields.Hostname
 	domain := routes[0].Spec.RouteSpecFields.Domain
-	urlPath := routes[0].Spec.RouteSpecFields.Path
 	labels := MakeVirtualServiceLabels(routes[0].Spec.RouteSpecFields)
-
-	var (
-		ownerRefs []metav1.OwnerReference
-		appNames  []string
-	)
-	for _, route := range routes {
-		// Each route will own the VirtualService. Therefore none of them can be a
-		// controller.
-		ownerRef := *kmeta.NewControllerRef(route)
-		ownerRef.Controller = nil
-		ownerRef.BlockOwnerDeletion = nil
-		ownerRefs = append(ownerRefs, ownerRef)
-
-		// AppNames
-		if route.Spec.AppName != "" {
-			appNames = append(appNames, route.Spec.AppName)
-		}
-	}
 
 	hostDomain := domain
 	if hostname != "" {
 		hostDomain = hostname + "." + domain
 	}
 
-	httpRoute, err := buildHTTPRoute(namespace, urlPath, appNames)
-	if err != nil {
-		return nil, err
+	var (
+		ownerRefs  []metav1.OwnerReference
+		appNames   []string
+		httpRoutes []networking.HTTPRoute
+	)
+	for _, route := range routes {
+		// Each route will own the VirtualService. Therefore none of them can
+		// be a controller.
+		ownerRef := *kmeta.NewControllerRef(route)
+		ownerRef.Controller = nil
+		ownerRef.BlockOwnerDeletion = nil
+		ownerRefs = append(ownerRefs, ownerRef)
+		urlPath := route.Spec.RouteSpecFields.Path
+
+		// AppNames
+		if route.Spec.AppName != "" {
+			appNames = append(appNames, route.Spec.AppName)
+		}
+
+		httpRoute, err := buildHTTPRoute(namespace, urlPath, appNames)
+		if err != nil {
+			return nil, err
+		}
+
+		httpRoutes = algorithms.Merge(
+			v1alpha1.HTTPRoutes(httpRoutes),
+			v1alpha1.HTTPRoutes(httpRoute),
+		).(v1alpha1.HTTPRoutes)
 	}
 
 	return &networking.VirtualService{
@@ -99,10 +105,7 @@ func MakeVirtualService(routes []*v1alpha1.Route) (*networking.VirtualService, e
 			),
 			Namespace:       v1alpha1.KfNamespace,
 			OwnerReferences: ownerRefs,
-			Labels: resources.UnionMaps(
-				labels, map[string]string{
-					ManagedByLabel: "kf",
-				}),
+			Labels:          labels,
 			Annotations: map[string]string{
 				"domain":   domain,
 				"hostname": hostname,
@@ -112,7 +115,7 @@ func MakeVirtualService(routes []*v1alpha1.Route) (*networking.VirtualService, e
 		Spec: networking.VirtualServiceSpec{
 			Gateways: []string{KnativeIngressGateway},
 			Hosts:    []string{hostDomain},
-			HTTP:     httpRoute,
+			HTTP:     httpRoutes,
 		},
 	}, nil
 }
@@ -167,6 +170,11 @@ func buildHTTPRoute(namespace, urlPath string, appNames []string) ([]networking.
 }
 
 func buildPathRegex(path string) (string, error) {
+	// If its just the root path, we'll add that back as optional
+	if path == "/" {
+		path = ""
+	}
+
 	p, err := (&mux.Router{}).PathPrefix(path).GetPathRegexp()
 	if err != nil {
 		return "", err

--- a/pkg/reconciler/route/resources/virtual_service_test.go
+++ b/pkg/reconciler/route/resources/virtual_service_test.go
@@ -18,11 +18,9 @@ import (
 	"errors"
 	"fmt"
 	"net/http"
-	"sort"
 	"testing"
 
 	"github.com/google/kf/pkg/apis/kf/v1alpha1"
-	"github.com/google/kf/pkg/kf/algorithms"
 	"github.com/google/kf/pkg/kf/testutil"
 	"github.com/google/kf/pkg/reconciler/route/resources"
 	"github.com/knative/serving/pkg/network"
@@ -242,7 +240,15 @@ func TestMakeVirtualService(t *testing.T) {
 }
 
 func ExampleMakeVirtualService() {
-	vs1, err := resources.MakeVirtualService([]*v1alpha1.Route{
+	vs, err := resources.MakeVirtualService([]*v1alpha1.Route{
+		{
+			Spec: v1alpha1.RouteSpec{
+				RouteSpecFields: v1alpha1.RouteSpecFields{
+					Hostname: "some-host",
+					Domain:   "example.com/",
+				},
+			},
+		},
 		{
 			Spec: v1alpha1.RouteSpec{
 				RouteSpecFields: v1alpha1.RouteSpecFields{
@@ -252,12 +258,6 @@ func ExampleMakeVirtualService() {
 				},
 			},
 		},
-	})
-	if err != nil {
-		panic(err)
-	}
-
-	vs2, err := resources.MakeVirtualService([]*v1alpha1.Route{
 		{
 			Spec: v1alpha1.RouteSpec{
 				RouteSpecFields: v1alpha1.RouteSpecFields{
@@ -272,18 +272,11 @@ func ExampleMakeVirtualService() {
 		panic(err)
 	}
 
-	r := algorithms.Merge(
-		v1alpha1.HTTPRoutes(vs1.Spec.HTTP),
-		v1alpha1.HTTPRoutes(vs2.Spec.HTTP),
-	).(v1alpha1.HTTPRoutes)
-
-	// Sort for display purposes
-	sort.Sort(r)
-
-	for i, h := range r {
+	for i, h := range vs.Spec.HTTP {
 		fmt.Printf("Regex %d: %s\n", i, h.Match[0].URI.Regex)
 	}
 
-	// Output: Regex 0: ^/some-path-1(/.*)?
-	// Regex 1: ^/some-path-2(/.*)?
+	// Output: Regex 0: ^(/.*)?
+	// Regex 1: ^/some-path-1(/.*)?
+	// Regex 2: ^/some-path-2(/.*)?
 }

